sql/pipelines/src/main/scala/org/apache/spark/sql/pipelines/logging/ConstructPipelineEvent.scala

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.pipelines.logging
+
+import java.sql.Timestamp
+import java.time.Instant
+import java.util.UUID
+
+import scala.util.control.NoStackTrace
+
+import org.apache.spark.internal.Logging
+
+case class UnresolvedFlowFailureException(name: String, cause: Throwable)
+    extends Exception(s"Failed to resolve flow '$name'", cause)
+    with NoStackTrace
+
+/**
+ * A wrapper for a serialized exception. When serialized, this wrapper object returns the original
+ * serialized exception.
+ */
+class SerializedExceptionWrapper(val exceptions: Seq[SerializedException])
+  extends Exception(exceptions.headOption.map(_.message).orNull)
+    with NoStackTrace
+
+/**
+ * A factory object that is used to construct [[PipelineEvent]]s with common fields
+ * automatically filled in. Developers should always use this factory rather than construct
+ * an event directly from an empty proto.
+ */
+object ConstructPipelineEvent extends Logging {
+
+  /**
+   * Mix-in trait that enables exceptions to have custom class name when recorded to the event
+   * buffer i.e. a class name that is different from the actual runtime class name of the exception.
+   */
+  private[pipelines] def serializeException(t: Throwable): Seq[SerializedException] = {
+    t match {
+      case serializedExceptionWrapper: SerializedExceptionWrapper =>
+        serializedExceptionWrapper.exceptions
+      case _ =>
+        val className = t.getClass.getName
+        t match {
+          case _: UnresolvedFlowFailureException =>
```

Comment:
why converting `UnresolvedFlowFailureException`  to `SerializedException`?

Reply:
`SerializedException` is a slightly more structured format for recording errors that happen during pipeline execution. We will eventually send these structured errors back to the client so customers can debug issues in their pipeline.